nilearn/glm/second_level/second_level.py

```diff
@@ -30,79 +30,82 @@
 from nilearn.glm._base import BaseGLM
 
 
-def _check_second_level_input(second_level_input, design_matrix,
-                              confounds=None, flm_object=True, df_object=True):
+def _check_second_level_input(second_level_input,
+                              design_matrix,
+                              confounds=None):
     """Checking second_level_input type"""
     # Check parameters
     # check first level input
-    if isinstance(second_level_input, list):
+
```

Comment:
I find this function long and difficult to understand. I would break the different pieces into (almost) single responsibility functions. Maybe something more or less like this:

```python
def _check_second_level_input(second_level_input, design_matrix, confounds=None):
    input_type = _check_input_type(second_level_input)
    _check_input_as_type(
        second_level_input, input_type, confounds is None, design_matrix is None
    )

def _check_input_type(second_level_input) -> str:
    """Determines the type of input provided."""
    if isinstance(second_level_input, pd.DataFrame):
         return "df_object"
    if isinstance(second_level_input, (str, Nifti1Image)):
         return "nii_object"
    if isinstance(second_level_input, list):
        return _check_input_type_when_list(second_level_input)
    raise TypeError("....")

def _check_input_type_when_list(second_level_input: listl) -> str:
    """Determines the type of input provided when it is a list."""
    if len(second_level_input) < 2:
        raise TypeError("...")
    _check_all_elements_of_same_type(second_level_input)   
    if all(isinstance(x, (str, Nifti1Image)) for x in second_level_input):
        return "nii_object"
    if all(isinstance(x, FirstLevelModel) for x in second_level_input):
        return "flm_object"
    raise TypeError("...")

def _check_all_elements_of_same_type(data: list) -> None:
    for idx, input in enumerate(data):
        if not isinstance(input, type(data[0])):  
             raise TypeError(
                 f'Elements of second_level_input must be of the same type.'
                 f' Got object type {type(input)} at idx {idx}.')

def _check_input_as_type(second_level_input, input_type, none_confounds: bool, none_design_matrix: bool):
    if input_type == "flm_object":
        _check_input_as_first_level_model(second_level_input, none_confounds)
    elif input_type == "nii_object":
         _check_input_as_nifti_images(second_level_input, none_design_matrix)
    else:
         _check_input_as_dataframe(second_level_input)

def _check_input_as_first_level_model(second_level_input, none_confounds: bool):
    for model_idx, first_level in enumerate(second_level_input):
        if (first_level.labels_ is None or first_level.results_ is None):
             raise ValueError(
                 'Model %s at index %i has not been fit yet'
                 '' % (first_level.subject_label, model_idx))
         if not none_confounds and first_level.subject_label is None:
             raise ValueError("...")

def _check_input_as_dataframe(second_level_input):
    for col in ('subject_label', 'map_name', 'effects_map_path'):
        if col not in second_level_input.columns:
            raise ValueError("...")
    if not all(isinstance(_, str) for _ in second_level_input['subject_label'].tolist()):
        raise ValueError('...')

def _check_input_as_nifti_images(second_level_input, none_design_matrix):
    if isinstance(second_level_input, (str, Nifti1Image)):
        second_level_input = [second_level_input]
    for niimg in second_level_input:
        check_niimg(niimg=niimg, ensure_ndim=4)
    if none_design_matrix:
         raise ValueError("...")
```

Reply:
I did have it flagged as one of the function that was too complex and needed refactoring so I am totally in.

Let me incorporate that into the PR.